src/axom/core/Array.hpp

```diff
@@ -1227,6 +1229,15 @@ inline void Array<T, DIM>::updateNumElements(IndexType new_num_elements)
 {
   assert(new_num_elements >= 0);
   assert(new_num_elements <= m_capacity);
+
+  int new_elems = new_num_elements - m_num_elements;
+  for(int ielem = 0; ielem < new_elems; ielem++)
+  {
+    // TODO: what to do here when T isn't default-constructible?
+    // we should probably do what std::vector does (zero out memory?)
+    new(&m_data[ielem + m_num_elements]) T;
```

Comment:
For what it's worth the docstring for `axom::Array` specifically calls out the non-default-construction of its data as being one of its main differences relative to `std::vector`.  That callout dates from before I started working on the class so I can't be sure of its motivation, though.

Reply:
I believe leaving the memory uninitialized was problematic for `axom::Array`s of non-trivial type. I think in my case, it was `axom::Array<BitSet>` - you can end up with the destructor for `BitSet` getting called on junk memory, which causes a segfault.

Although I understand the desire to avoid doing unnecessary initializations - maybe we can only do placement-new for cases where `std::is_trivially_destructible<T>` is false.

Reply:
Yep - I think what I'll end up doing is adding a `bool` member that allows the user to select at construction time whether they want data to be initialized to `T{}` when it's allocated - possibly via a new set of constructors that take a "dummy"/sentinel param:

```cpp
// Technically we could add a boolean arg to the 1D constructor instead
Array<int> foo(axom::ArrayOptions::Uninitialized{}, N);
// but this would be much more annoying for the multidimensional case, and I'd prefer to keep things consistent
Array<int, 3> bar(axom::ArrayOptions::Uninitialized{}, M, N, K);
```

That being said I have no problem with having the Array changes in this PR merged as-is for convenience/expediency.